text/0000-async_await.md

```diff
@@ -0,0 +1,571 @@
+Prior to 1.0, we experimented with having a green-threading runtime built into
+the language. However, this proved too opinionated - because it impacted every
+program written in Rust - and it was removed shortly before 1.0. After 1.0,
+asynchronous IO initially focused around the mio library, which provided an
+cross-platform abstraction over the async IO primitives of Linux, Mac OS, and
+Windows. In mid-2016, the introduction of the futures crate had a major impact
+by providing a convenient, shared abstraction for asynchronous operations. The
+tokio library provided a mio-based event loop that could execute code
+implemented using the futures interfaces.
+
+After gaining experience & user feedback with the futures-based ecosystem, we
+discovered certain ergonomics challenges. Using state which needs to be shared
+across await points was extremely ergonomic - requiring either Arcs or join
+chaining - and while combinators were often more ergonomic than manually
+writing a future, they still often led to messy sets of nested and chained
+callbacks.
+
+Fortunately, the Future abstraction is well suited to use with a syntactic
+sugar which has become common in many languages with async IO - the async and
+await keywords. In brief, an asynchronous function returns a future, rather
+than evaluating immediately when it is called. Inside the function, other
+futures can be awaited using an await expression, which causes them to yield
+control while the future is being polled. From a user's perspective, they can
+use async/await as if it were synchronous code, and only need to annotate their
+functions and calls.
+
+Async/await & futures can be a powerful abstraction for asynchronicity and
+concurrency in general, and likely has applications outside of the asynchronous
+IO space. The use cases we've experience with today are generally tied to async
+IO, but by introducing first class syntax and libstd support we believe more
+use cases for async & await will also flourish, that are not tied directly to
+asynchronous IO.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Async functions
+
+Functions can be annotated with the `async` keyword, making them "async
+functions":
+
+```rust
+async fn function(argument: &str) -> usize {
+     // ...
+}
+```
+
+Async functions work differently from normal functions. When an async function
+is called, it does not enter the body immediately. Instead, it evaluates to an
+anonymous type which implements `Future`. As that future is polled, the
+function is evaluated up to the next `await` or return point inside of it (see
+the await syntax section next).
+
+An async function is a kind of delayed computation - nothing in the body of the
+function actually runs until you begin polling the future returned by the
+function. For example:
+
+```rust
+async fn print_async() {
+     println!("Hello from print_async")
+}
+
+fn main() {
+     let future = print_async();
```

Comment:
I'm worried this syntax obfuscates what's actually occurring when you evaluate the code that binds future. Currently you only need to declare types in function parameters and returns. However async functions are now getting ergonomic sugar so you don't need to return a `Future<T>`. In exchange we should add verbosity elsewhere to clarify the initial call to print_async() does nothing. Some examples:

`let future: Future = print_async();`
`let async future = print_async();`   
`let future: async = print_async();`

Reply:
(See previous discussion in this thread around "implicit await" and /u/munificient's comments, which addresses that confusion by making the default case immediately await and only allows deferred execution with an explicit annotation.)

Reply:
In keeping with Rust's usual explicitness, I'd rather *neither* case were implicit. I'd be happier in a world where something along the lines of `let future = async!(print_async());` indicates the case where one does not greedily evaluate to the first await, while `let future = await!(print_async())` indicates the case where one does. That way there is no way to confuse async/await with regular function invocation.

Reply:
Merely being explicit is insufficient justification for any decision. There are many implicit things in Rust, most of them crucial for writing readable code.

I disagree that it implicit await obfuscates anything. It makes control flow consistent with sync functions- a call immediately runs the body of the callee, and an async-block-as-value/closure-as-value defers it.

The problem of knowing whether you can trust some shared state to remain valid between two points in a function is not solved merely by making awaits explicit- even a sync function call can be a problem! In that sense, an await is basically just a call to a rather large function.

Reply:
The difference is, with a function, even a large one, the code being called is deterministic. With async it is not. And there are other differences. With a function call, it either returns, panics or the application terminates (or runs forever), the three later ones probably a bug. It can be also dropped and never resumed even in the „correct“ case with async function.

Reply:
The code a function runs is no more deterministic than an await. It might go through a trait object, it might pull some work off a queue, and for that matter Rust is Turing complete anyway.

The possibilities of "return, panic, abort, or hang" are also effectively the same. Returning corresponds to the async function being resumed; panicing corresponds to it being dropped; and abort and hang are identical.

So a function call can do *quite literally* anything an await can do, and it can start or stop doing it without forcing a breaking change. Nothing about an explicit `await` marker prevents this, unless you also give up on synchronous function calls.

Fortunately, *this is not a big deal* in either sync *or* async Rust- we already have the borrow checker and `Sync`/`Send` systems to prevent state from getting changed out from under you. Interior mutability is already explicitly marked. `RefCell`/`RwLock` already give you even finer-grained dynamic control over shared mutable state. If you're still worried about this problem the solution is to fix it directly, not to sprinkle annotations in async code and hope people pay attention.

Reply:
> So a function call can do quite literally anything an await can do

No it can't: code can't safely early-return from a function further up the callstack without either panicking or aborting the program entirely.

> Fortunately, this is not a big deal in either sync or async Rust

Adding this would be a huge deal - the only "implicit" form of early return that rust has is panicking, and that's already bad enough: it's already considered part of the "contract" of a function whether it can panic or not, and making a function previously documented as "not panicking" start panicking is a clear breaking change.

Making yields implicit adds a whole load of foot-guns to the language for very little reason: it's very easy to write "async-safe" code when yields are explicit: it's a lot harder when any function you call could possibly yield. Am I now supposed to audit every function I use in case it might yield? That's not acceptable to me, and failing to do that results in one of the kind of errors that rust's ownership model can't actually help you with: unforeseen and untested inter-leavings of async execution which result in bugs.

Luckily with panics you only really have to worry about them when writing unsafe code, and so this "auditing" process is a bit simpler, but there's no marker like "unsafe" for code which might have "bad inter-leavings".

Reply:
You misunderstand. A function doesn't need to "safely early-return" to introduce "bad interleaving." It can do that all on its own, simply by performing an action you didn't want it to, or by recursively calling a function from higher in the call stack.

The point about "forcing a breaking change" was a reference to @vorner's terminology- earlier they mentioned wanting the change `fn foo` -> `async fn foo` to prevent its call sites from compiling- a *forced* breaking change. Your use of the term "breaking change" is more general- and it can be met by almost *any* change, not just introducing a panic.

So yes, you are "now supposed to audit every function you use in case it might yield"... but only because you *already* have to "audit" every function you use it case it might do something other than what you want.